### DISCUSSION TITLE
Issue running 64 bit Windows AIR Application

### DISCUSSION THREAD

**Discussion by rogerwilliam1990** [airsdk/Adobe-Runtime-Support#1824] [Q&A]
Using the latest SDK (v33.1) we complied a 64 bit AIR application by adding 

- Adding `<architecture>64</architecture> ` in the application descriptors
- Build using "-arch x64"

When we try to run the application we are getting the following error 

**_"Initial content could not be loaded for this application. Try reinstalling or contact the published for assistance"_**

We also tried using the sample code provided along SDK (AIRSDK_Windows\samples) we are getting the same error when we try to run the application.

Are we missing something? Kindly help. Thanks!

**Comment by ajwfrost** [comment]
This is odd, it works for our test cases here. I'm wondering if there may be an anti-virus application blocking the different DLL from loading.... but can you please try using ADL to start with (adl64.exe) i.e. `sdkfolder\bin\adl64 application.xml swf_folder`

If you build a package, is this using `-target bundle` or `-target native`? If you do the 'bundle' you should end up with a folder containing an executable file and your SWF, plus a subfolder "Adobe AIR" that contains the AIR runtime DLL. Does that all look correct?

And just checking that you're not on Windows 11..?

thanks

**Reply by rogerwilliam1990** [reply]
We did try running the application using adl64.exe the outcome is same. 

Now if I build using `-target native` the outcome is an .exe installer and its a 32bit application and it works fine without any issues. But we want to compile as 64 bit application, for that we had to use `-target bundle`, the issue pops up only at this time.

We are using Windows 10. We also had that doubt regarding Anti Virus. So we tried creating an Virtual Machine and it only had AIR SDK, still we face this issue.

Thanks!

**Reply by ajwfrost** [reply]
So with adl64.exe, you are still getting a message "Initial content could not be loaded for this application"?
Do you have any native extensions here?

The odd thing is that this all implies that the Adobe AIR DLL is being loaded into the system, otherwise you shouldn't see that error (it would just say something like "Can't find the AIR runtime, please add -runtime to the ADL command line"...

**Reply by rogerwilliam1990** [reply]
Yes when I run the application with adl64.exe I get **_"error while loading initial content"_** on the command line and the application crashes/closes and fails to launch

When I run the application by double clicking I get a popup with error **_"Initial content could not be loaded for this application. Try reinstalling or contact the published for assistance"_**

The behavious is same with our code and also with sample code provided along SDK (AIRSDK_Windows\samples)

Thanks!

**Reply by rogerwilliam1990** [reply]
@ajwfrost Kindly note that we are using HTML and not SWF. So we specify the HTML file in the `<content>` node of the descriptor file

**Reply by ajwfrost** [reply]
I've just tried with the sample app descriptor from the SDK "samples" folder, after changing the SWF filename and the 'visible' property to 'true', it worked fine with adl64.exe here. I'm just thinking, the "loading initial content" may mean that it's got a problem loading in the SWF file itself, or initializing it (like any exception thrown during the main class initializer may cause this problem).

So can you try with a different/blank SWF application? Or could we get your SWF file to have a look at what may be causing this?

thanks

**Reply by rogerwilliam1990** [reply]
Thank you, but our application is based on HTML and not SWF. We tried using "AIRSDK_Windows\samples\badge\red_badge.html" when error came up.

We also tried a blank HTML with minimal content, faced the same issue.

Could you please advice on loading HTML content. 

Thanks!

**Reply by ajwfrost** [reply]
Ah ... so that would explain it, I should have thought of that!

Let me check on this...

thanks

**Comment by 2jfw** [comment]
Can you check and verify in your _compiled_ folder that in the folder `META-INF\AIR` in the file `application.xml` the `content`-tag under `initialWindow` does actually point to an existing file?

`	<!-- Settings for the application's initial window. Required. -->`
`	<initialWindow>`
`		<!-- Note: In Flash Builder, the SWF reference is set automatically. -->`
`		<!--   HINT: THIS IS WHEN STARTING DEBUGGER WITHOUT INVOKE ARGUMENTS! -->`
	`	<content>DoesThisFileExistRelativeToRootQuestionmark</content>`

**Reply by concord232** [reply]
Yes @2jfw  it does point to the existing html file.

Please note the same source works for 32 bit Native. Just that it does not work for 64 bit Bundle

thanks

**Reply by 2jfw** [reply]
HTML file you say? Mm, can you change that value to point to your (Main)SWF instead?

**Reply by concord232** [reply]
@2jfw Ours is a HTML based application and not SWF. The content points to our main HTML. 

Our application is similar to as in this post -> https://github.com/airsdk/Adobe-Runtime-Support/discussions/1807#discussioncomment-2586720

**Reply by 2jfw** [reply]
The thing is, there is **no** SWF file!
AFAIK the EXE is just calling the SWF - in your case its calling an HTML file and this does not seem right to me.
You need to build an SWF file and link it in the `initialWindow/content` tag

**EDIT:** Thanks for the link - I now see what you are trying to achieve - never came around that method of linking content/PDF so please ignore my comment

**Reply by concord232** [reply]
@2jfw The initialWindow/content can be HTML kindly refer to following link for details

https://help.adobe.com/en_US/air/build/WSfffb011ac560372f2fea1812938a6e463-8000.html#WSfffb011ac560372f2fea1812938a6e463-7ff4

Thanks!

**Reply by ajwfrost** [reply]
I just realised we'd not updated this thread, we'd looked into this. When you specify an HTML file for the `initialWindow.content` value, the AIR runtime is actually wrapping that into a SWF file that's based on HTMLLoader. This wrapping code was omitted when we took out the original HTMLLoader implementation and added the updated version based on StageWebView, just because of the way the feature flags had been originally set up.

We've updated this SWF wrapper so that the HTML content is loaded into a StageWebView and then it starts up like normal - but there are still differences between the original HTMLLoader functionality vs the StageWebView based version. So it might be that some applications need tweaking to make the HTML/JS code work properly...

**Reply by concord232** [reply]
@ajwfrost Thanks! 

When can we expect the release of SDK with this fix?
Kindly also advice on the tweakings needed for HTML/JS code to work properly.

**Reply by ajwfrost** [reply]
Normally I'd expect the next release to be around 9th May - we were trying to get into a pattern of cutting a release candidate at the start of each month.. 

In terms of the changes needed, it very much depends on what is there currently! On Windows though: if the user doesn't have the Edge "WebView2" component installed, then it will be using the IE-based "WebBrowser" component so it may be worth checking that your HTML/JS is all compatible with IE. The easiest thing to do here would be to create a SWF-based AIR application and use this to load in your HTML code to a StageWebView object, and see what happens! We can provide a suitable SWF if necessary which would just load a named HTML file upon start-up similar to how this will work once released..

thanks

**Reply by concord232** [reply]
@ajwfrost Thanks for the feedback. 

But one thing I am not able to understand is why 32 bit Native application is working on the same PC and 33.1 SDK but a 32bit bundle or 64 bit bundle fails to work?

**Reply by ajwfrost** [reply]
"Native" means we build an installer file which then uses the shared AIR runtime, whatever is on the computer (although it also checks to see whether a newer version is required, but that's a different issue..). "Bundle" will use the captive runtime from the SDK and embed this into the application bundle.

So probably if you're seeing different behaviour, then you have a shared/common AIR runtime on the computer which is a lower version than the AIR SDK that you're using. You can check that via Add/Remove Programs, search for Adobe AIR and check its version. Or look in "c:\Program Files (x86)\Common Files\Adobe AIR\Versions\1.0" and check the file properties/details of the DLL.

thanks

**Reply by concord232** [reply]
@ajwfrost  Thanks that explains :-) 

32 Native application is using version 33.1.1.743 which is with custiom build of webkit according to online source https://airsdk.harman.com/runtime

But 32/64 bit bundle application are created using AIR SDK version 33.1.1.821 which I suppose does not include the same.

Can I get a version of AIR SDK which has runtimes with custiom build of webkit
or alternatively can I get version 33.1.1.743 SDK?

Thanks!

**Reply by ajwfrost** [reply]
Yes, so the 33.1.1.743 versions will have the old webkit library. We've left these available for download as a shared runtime (at the bottom of that https://airsdk.harman.com/runtime page) although the version at the top of the page is now 33.1.1.821.

The SDK versions you can get by clicking on the appropriate version number from https://airsdk.harman.com/release_notes, so if you want the 33.1.1.743 SDK then it would take you to https://airsdk.harman.com/download/33.1.1.743

There's not really any mixing and matching i.e. you can't have the latest SDK but have this use runtimes that are older. We had been wondering about reverting things so that we re-include the webkit libraries in the 33.1 branch, but it's not really sustainable so I think if people need that library they'll need to continue with the 33.1.1.743 version. If there are any critical/showstopper errors that folk are finding with 33.1.1.743 then we could perhaps do an update of it but otherwise, we'll continue with the below approaches:

1) if you need the original HTMLLoader/webkit functionality, you can use 33.1.1.743 SDK and/or runtimes
2) we'll work on things like this (to re-enable HTML files as the main content file) to try to help with compatibility and to encourage people to switch to later SDKs/runtimes and have their apps work with this
3) we're looking at what AS/JS interaction people were using, to see if we can provide a way to enable this via a JS library
4) we're also looking at ways we could offer an integration route to this for third-party browser engines - which is more generic really, i.e. a way for third party components to draw their contents into the AIR display list.



thanks

**Reply by concord232** [reply]
@ajwfrost  Thanks! We are ok to use 33.1.1.743 version SDK for now. as long as html loading and PDF render works.

With Sample application https://github.com/airsdk/Adobe-Runtime-Support/discussions/1807#discussioncomment-2586720 I tried creating 32 bit and 64 bit bundle application using 33.1.1.743 version SDK. The application launches now without any problem.

32 bit bundle application renders PDF without any issues. (32 bit Adobe Reader is Installed)
But 64 bit bundle application fails to render PDF. (64 bit Adobe Reader is installed)

Kindly help in this regards.

**Comment by mesaglam** [comment]
Hi, I've built the app with captive runtime. But it doesn't work usb device's root folder. I'm getting same error "Initial content could not be loaded for this application. Try reinstalling or contact the published for assistance"

But if i create a folder and put the files inside in it, then it works. What do you think?

here 

![pic](https://github.com/airsdk/Adobe-Runtime-Support/assets/7021706/1c544dcf-e98b-4d42-9b53-3499c332cf12)

**Reply by itlancer** [reply]
This is old AIR bug:
https://tracker.adobe.com/#/view/AIR-3940278
https://community.adobe.com/t5/air-discussions/windows-file-url-throw-error-when-application-executable-in-root-directory/td-p/6830964#7195184

**Reply by mesaglam** [reply]
And it still exists for 5 years? Interesting ...

**Reply by ajwfrost** [reply]
Hi - I'm not 100% sure whether these are the same though.. The old bug referenced above allows the application to load, and then throws an error when you try to access a File (turns out this is caused by the fact Windows keeps the backslash on the root path, and the AS3 code doesn't handle this .. fairly easy fix in the end!)

But @mesaglam your issue is a bit more fundamental, if it can't actually find the initial SWF file. Just out of interest, in your application descriptor file, do you have just:
`<content>Eysis.swf</content>`
or is there anything else there?

Or perhaps, do you have any ANEs? this might trigger the same problem I guess...

thanks

**Reply by ajwfrost** [reply]
Okay thanks - bit odd then, I can't reproduce it with a simple test case running on a `D:\` memory stick. But hopefully this is the same problem as we have when using `File.applicationDirectory.resolvePath()` per the existing Adobe bug report. We'll put this fix into our next release so please check again with that (in 2-3 weeks hopefully)

thanks

**Reply by mesaglam** [reply]
Yes, it was because of ANE. After removing the ane files everything worked. Sorry to bother you.

**Reply by ajwfrost** [reply]
Ah so yes, probably when it loads in an ANE it's using this same mechanism to resolve a file path, so it's very likely that this was the same problem. Should be resolved in 50.2.3.1...

**Reply by mesaglam** [reply]
Thanks, you rock!
